wincrypto/src/dtls.rs

```diff
@@ -0,0 +1,601 @@
+                self.security_ctx.as_ref().unwrap() as *const _,
+                SECPKG_ATTR_STREAM_SIZES,
+                &mut self.encrypt_message_input_sizes as *mut _ as *mut std::ffi::c_void,
+            )
+            .map_err(|e| WinCryptoError(format!("SECPKG_ATTR_STREAM_SIZES: {:?}", e)))?;
+
+            let mut srtp_parameters = SecPkgContext_SrtpParameters::default();
+            QueryContextAttributesW(
+                self.security_ctx.as_ref().unwrap() as *const _,
+                SECPKG_ATTR(SECPKG_ATTR_SRTP_PARAMETERS),
+                &mut srtp_parameters as *mut _ as *mut std::ffi::c_void,
+            )
+            .map_err(|e| {
+                WinCryptoError(format!("QueryContextAttributesW Keying Material: {:?}", e))
+            })?;
+
+            let srtp_profile_id = u16::from_be(srtp_parameters.ProtectionProfile);
+
+            let keying_material_info = SecPkgContext_KeyingMaterialInfo {
+                cbLabel: DTLS_KEY_LABEL.len() as u16,
+                pszLabel: windows::core::PSTR(DTLS_KEY_LABEL.as_ptr() as *mut u8),
+                cbKeyingMaterial: srtp_keying_material_len(srtp_profile_id)?,
+                cbContextValue: 0,
+                pbContextValue: std::ptr::null_mut(),
+            };
+            SetContextAttributesW(
+                self.security_ctx.as_ref().unwrap() as *const _,
+                SECPKG_ATTR_KEYING_MATERIAL_INFO,
+                &keying_material_info as *const _ as *const std::ffi::c_void,
+                std::mem::size_of::<SecPkgContext_KeyingMaterialInfo>() as u32,
+            )
+            .map_err(|e| {
+                WinCryptoError(format!("SetContextAttributesA Keying Material: {:?}", e))
+            })?;
+
+            let mut keying_material = SecPkgContext_KeyingMaterial::default();
+            QueryContextAttributesExW(
+                self.security_ctx.as_ref().unwrap() as *const _,
+                SECPKG_ATTR(SECPKG_ATTR_KEYING_MATERIAL),
+                &mut keying_material as *mut _ as *mut std::ffi::c_void,
+                std::mem::size_of::<SecPkgContext_KeyingMaterial>() as u32,
+            )
+            .map_err(|e| {
+                WinCryptoError(format!(
+                    "QueryContextAttributesExW Keying Material: {:?}",
+                    e
+                ))
+            })?;
+
+            let srtp_keying_material = std::slice::from_raw_parts(
+                keying_material.pbKeyingMaterial,
+                keying_material.cbKeyingMaterial as usize,
+            )
+            .to_vec();
+
+            FreeContextBuffer(keying_material.pbKeyingMaterial as *mut _ as *mut std::ffi::c_void)
+                .map_err(|e| {
+                    WinCryptoError(format!("FreeContextBuffer Keying Material: {:?}", e))
+                })?;
+
+            let mut peer_cert_context: *mut CERT_CONTEXT = std::ptr::null_mut();
+            QueryContextAttributesW(
+                self.security_ctx.as_ref().unwrap() as *const _,
```

Comment:
return error instead of unwrap

Reply:
or if you are expected to have a ctx then use expect with comment why it's ok

Reply:
So.. changed it to return an Err. I tried to make the str0m_wincrypto always return Result<> for its functions. The str0m mod should "expect" if it doesn't anticipate failures.